pkgs/build-support/docker/default.nix

```diff
@@ -866,4 +868,77 @@ rec {
         makeWrapper ${streamScript} $out --add-flags ${conf}
       '';
     in result;
+
+  # This function builds a docker image that behaves like a nix-shell
+  buildNixShellImage =
+    # Use the environment of this derivation
+    { name ? drv.name + "-env"
+    , drv
+    , tag ? null
+    # Files to put on the image (a nix store path or list of paths).
+    , contents ? []
+    # uid/gid to run as inside the container
+    # Some compilers don't like to run as root
+    , uid ? 1000, gid ? 1000
+  }:
+    assert lib.assertMsg (! (drv.__structuredAttrs or false))
+      "buildNixShellImage does not work with the derivation ${drv.drvPath} because it uses __structuredAttrs";
+    let
+
+      setupPackage = runCommandNoCC "setup" {
+        profile = ''
+          # This mimics part of nix-shell's initialization
+          source $stdenv/setup
+          set +e
+          cd /home/user
+          runHook shellHook
+        '';
+        passAsFile = [ "profile" ];
+      } ''
+        mkdir -p $out/{etc/ssl/certs,tmp,bin,usr/bin,home/user/build}
+        ln -s ${bashInteractive}/bin/sh $out/bin/sh
+        ln -s ${coreutils}/bin/env $out/usr/bin/env
+        cp "$profilePath" $out/etc/profile
+        ln -s "${cacert}/etc/ssl/certs/ca-bundle.crt" $out/etc/ssl/certs/ca-certificates.crt
+      '';
+
+      # A mapping from output name to the nix store path where they should end up
+      outputPaths = lib.genAttrs drv.outputs (output: builtins.unsafeDiscardStringContext drv.${output}.outPath);
```

Comment:
Are these paths writable in the store?
We should have a test case for this as well.

It would be nice to have symlinks to these paths in a fixed location for derived images to use.

Reply:
They are indeed writable. Perhaps I could create /home/user/outputs/* again, symlinking into the store

Reply:
Now created symlinks from `/home/user/outputs/*` into the `/nix/store` paths